lib/spear/connection.ex

```diff
@@ -115,22 +116,94 @@ defmodule Spear.Connection do
   defp process_response(_unknown, state), do: state
 
   defp stream_body(state, request_ref, messages) do
-    Enum.reduce_while(messages, {:ok, state}, &stream_body_message(&1, &2, request_ref))
+    Enum.reduce_while(
+      messages,
+      {:ok, state},
+      &stream_body_message(&1, &2, request_ref)
+    )
   end
 
   defp stream_body_message(message, {:ok, state}, request_ref) do
-    {wire_data, _byte_size} = Request.to_wire_data(message)
-
-    stream_result =
-      Mint.HTTP.stream_request_body(
-        state.conn,
-        request_ref,
-        wire_data
-      )
-
-    case stream_result do
-      {:ok, conn} -> {:cont, {:ok, put_in(state.conn, conn)}}
-      {:error, conn, reason} -> {:halt, {:error, put_in(state.conn, conn), reason}}
+    {wire_data, byte_size} = Request.to_wire_data(message)
+    smallest_window = get_smallest_window(state.conn, request_ref)
+
+    with false <- byte_size > smallest_window,
+         {:ok, conn} <- Mint.HTTP.stream_request_body(state.conn, request_ref, wire_data) do
+      {:cont, {:ok, put_in(state.conn, conn)}}
+    else
+      _window_too_small? = true ->
+        recv_until_window_size_increase(state, smallest_window, request_ref)
+
+      {:error, conn, reason} ->
+        {:halt, {:error, put_in(state.conn, conn), reason}}
+    end
+  end
+
+  defp recv_until_window_size_increase(state, current_window, request_ref) do
+    params = {current_window, request_ref}
+
+    {:ok, state}
+    |> do_stage(:recv_responses, params)
+    |> do_stage(:check_window_size, params)
+    |> emit_stage_results()
+  end
+
+  defp do_stage({:error, state, reason}, _stage, _params), do: {:error, state, reason}
+
+  defp do_stage({:ok, state}, :recv_responses, {_current_window, _request_ref}) do
+    case receive_next_and_stream(state.conn) do
```

Comment:
initially this was written with `Mint.HTTP.recv/3`, but that function didn't seem to be increasing the connection's window size, and it seemed to be consistently timing out

I looked at how `recv/3` and `Mint.HTTP.stream/2` are implemented and I can't see any reason why `recv/3` would not adjust the connection window but /shrug `receive/1`+`stream/2` just works